### PR TITLE
refactor(identity): follow naming conventions for conversion methods

### PR DESCRIPTION
**Pull request by drHuangMHT**
## Description

This PR renames some method names that don't follow Rust naming conventions or behave differently from what the name suggests: 
- Enforce "try" prefix on all methods that return `Result`.
- Enforce "encode" method name for methods that return encoded bytes.
- Enforce "to_bytes" method name for methods that return raw bytes.
- Enforce "decode" method name for methods that convert encoded key.
- Enforce "from_bytes" method name for methods that convert raw bytes.

## Notes & open questions

Some method are intentionally left out to prevent conflict with #3681.  
~~Base branch is contaminated by #3621.~~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: identity/src/ecdsa.rs [hunk 1]
```diff
@@ -108,11 +108,22 @@ impl SecretKey {
...
 
     /// Decode a secret key from a byte buffer.
+    #[deprecated(
+        since = "0.2.0",
+        note = "This method name does not follow Rust naming conventions, use `SecretKey::try_from_bytes` instead"
+    )]
     pub fn from_bytes(buf: &[u8]) -> Result<Self, DecodingError> {
         SigningKey::from_bytes(buf)
             .map_err(|err| DecodingError::failed_to_parse("ecdsa p256 secret key", err))
             .map(SecretKey)
```
Comment (thomaseizinger) [inline identity/src/ecdsa.rs]: Should forward to the new function instead of duplicating code.

File: identity/src/ecdsa.rs [hunk 2]
```diff
@@ -158,10 +178,15 @@ impl PublicKey {
 
     /// Decode a public key into a DER encoded byte buffer as defined by SEC1 standard.
     pub fn decode_der(k: &[u8]) -> Result<PublicKey, DecodingError> {
```
Comment (thomaseizinger) [inline identity/src/ecdsa.rs:182]: Missing `deprecated` attribute.

File: identity/src/ed25519.rs [hunk 3]
```diff
@@ -164,12 +176,32 @@ impl PublicKey {
...
+    ///
+    #[deprecated(
+        since = "0.2.0",
+        note = "This method name does not follow Rust naming conventions, use `PublicKey::try_from_bytes` instead."
+    )]
     pub fn decode(k: &[u8]) -> Result<PublicKey, DecodingError> {
+        Self::try_from_bytes(k)
+    }
+
+    /// Decode a public key from a byte array as produced by `encode`.
```
Comment (thomaseizinger) [inline identity/src/ed25519.rs]:
```suggestion
    /// Decode a public key from a byte array as produced by `to_bytes`.
```

File: identity/src/ed25519.rs [hunk 4]
```diff
@@ -49,7 +49,19 @@ impl Keypair {
...
+    #[deprecated(
+        since = "0.2.0",
+        note = "This method name does not follow Rust naming conventions, use `Keypair::try_from_bytes` instead."
+    )]
     pub fn decode(kp: &mut [u8]) -> Result<Keypair, DecodingError> {
+        Self::try_from_bytes(kp)
+    }
+
+    /// Decode a keypair from the [binary format](https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.5)
+    /// produced by [`Keypair::encode`], zeroing the input on success.
```
Comment (thomaseizinger) [inline identity/src/ed25519.rs]:
```suggestion
    /// produced by [`Keypair::to_bytes`], zeroing the input on success.
```

File: identity/src/error.rs [hunk 5]
```diff
@@ -156,3 +158,26 @@ impl Error for SigningError {
         self.source.as_ref().map(|s| &**s as &dyn Error)
     }
 }
+
+/// Error produced when failing to convert `libp2p_identity::Keypair` to a more concrete keypair.
```
Comment (thomaseizinger) [inline identity/src/error.rs]:
```suggestion
/// Error produced when failing to convert [`Keypair`](crate::Keypair) to a more concrete keypair.
```

File: identity/src/error.rs [hunk 6]
```diff
@@ -156,3 +158,26 @@ impl Error for SigningError {
...
+#[derive(Debug)]
+pub struct OtherVariantError {
+    actual: KeyType,
+}
+
+impl OtherVariantError {
+    pub fn new(actual: KeyType) -> OtherVariantError {
```
Comment (thomaseizinger) [inline identity/src/error.rs]:
```suggestion
    pub(crate) fn new(actual: KeyType) -> OtherVariantError {
```

**Review by thomaseizinger — COMMENTED** [review]
Very nicely done, only a few nits!

Also, can you add a changelog entry to `libp2p-identity` please and bump the patch version in the manifest of `libp2p-identity`?

**Comment by thomaseizinger** [comment]
> Base branch is contaminated by #3621.

Fixed it for you :)

**Comment by drHuangMHT** [comment]
I've intentionally kept `Cargo.lock` untouched. Should I commit the change?

**Comment by thomaseizinger** [comment]
> I've intentionally kept `Cargo.lock` untouched. Should I commit the change?

Yes, the lockfile needs to be up to date.

File: identity/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,10 @@
+## 0.1.2
```
Comment (thomaseizinger) [inline identity/CHANGELOG.md]:
```suggestion
## 0.1.2 - unreleased
```

File: identity/src/error.rs [hunk 8]
```diff
@@ -156,3 +158,26 @@ impl Error for SigningError {
         self.source.as_ref().map(|s| &**s as &dyn Error)
     }
 }
+
+/// Error produced when failing to convert [`Keypair`](crate::Keypair) to a more concrete keypair.
+#[derive(Debug)]
+pub struct OtherVariantError {
```
Comment (thomaseizinger) [inline identity/src/error.rs:164]: The error module is private so we need to export this from the root.

**Review by thomaseizinger — COMMENTED** [review]
Nicely done, only one more things!

File: protocols/gossipsub/src/behaviour.rs [hunk 9]
```diff
@@ -203,7 +203,7 @@ impl From<MessageAuthenticity> for PublishConfig {
         match authenticity {
             MessageAuthenticity::Signed(keypair) => {
                 let public_key = keypair.public();
-                let key_enc = public_key.to_protobuf_encoding();
+                let key_enc = public_key.encode_protobuf();
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour.rs:206]: This crate now depends on at least `libp2p-identity 0.1.2`, can you update the Cargo.toml please?

File: protocols/identify/src/protocol.rs [hunk 10]
```diff
@@ -169,7 +169,7 @@ where
         .map(|addr| addr.to_vec())
         .collect();
 
-    let pubkey_bytes = info.public_key.to_protobuf_encoding();
+    let pubkey_bytes = info.public_key.encode_protobuf();
```
Comment (thomaseizinger) [inline protocols/identify/src/protocol.rs:172]: Same here.

File: transports/noise/src/io/handshake.rs [hunk 11]
```diff
@@ -214,7 +214,7 @@ where
     let pb = pb_result?;
 
     if !pb.identity_key.is_empty() {
-        let pk = identity::PublicKey::from_protobuf_encoding(&pb.identity_key)?;
+        let pk = identity::PublicKey::try_decode_protobuf(&pb.identity_key)?;
```
Comment (thomaseizinger) [inline transports/noise/src/io/handshake.rs:217]: And here.

File: transports/plaintext/src/handshake.rs [hunk 12]
```diff
@@ -54,7 +54,7 @@ impl HandshakeContext<Local> {
     fn new(config: PlainText2Config) -> Self {
         let exchange = Exchange {
             id: Some(config.local_public_key.to_peer_id().to_bytes()),
-            pubkey: Some(config.local_public_key.to_protobuf_encoding()),
+            pubkey: Some(config.local_public_key.encode_protobuf()),
```
Comment (thomaseizinger) [inline transports/plaintext/src/handshake.rs:57]: And here.

File: transports/tls/src/certificate.rs [hunk 13]
```diff
@@ -159,7 +159,7 @@ fn parse_unverified(der_input: &[u8]) -> Result<P2pCertificate, webpki::Error> {
             //    required KeyType Type = 1;
             //    required bytes Data = 2;
             // }
-            let public_key = identity::PublicKey::from_protobuf_encoding(&public_key)
+            let public_key = identity::PublicKey::try_decode_protobuf(&public_key)
```
Comment (thomaseizinger) [inline transports/tls/src/certificate.rs:162]: And here.

**Review by thomaseizinger — APPROVED** [review]
This is ready for merge. We are just missing some version bumps to point to the correct version of `libp2p-identity`.

@mxinden We are renaming several functions to be consistent with Rust's naming guidelines. Unfortunately, there is no clippy lint for this.

**Comment by thomaseizinger** [comment]
Awesome, thank you for sending this PR!

**Review by mxinden — COMMENTED** [review]
Sorry for the late review here. Great work @drHuangMHT!

**Pull request merged**